pyramid/tests/test_integration.py

```diff
@@ -712,7 +712,234 @@ def test_wildcards(self):
         self.assertEqual(res.content_type, 'application/json')
         res = self.testapp.get('/hello', headers={'Accept': 'text/*'}, status=200)
         self.assertEqual(res.content_type, 'text/plain')
+        res = self.testapp.get('/hello', headers={'Accept': '*/*'}, status=200)
+        self.assertEqual(res.content_type, 'text/plain')
+
+    def test_no_accept(self):
+        res = self.testapp.get('/hello', status=200)
+        self.assertEqual(res.content_type, 'text/plain')
+
+    def test_invalid_accept(self):
+        res = self.testapp.get('/hello', headers={'Accept': 'foo'}, status=200)
+        self.assertEqual(res.content_type, 'text/plain')
+
+class AddViewAcceptArgMediaRangeAllTest(unittest.TestCase):
+    def setUp(self):
+        def view(request):
+            return 'text/plain'
+        from pyramid.config import Configurator
+        config = Configurator()
+        config.add_route('root', '/')
+        config.add_view(
+            view, route_name='root', accept='*/*', renderer='string',
+        )
+        app = config.make_wsgi_app()
+        from webtest import TestApp
+        self.testapp = TestApp(app)
+
+    def tearDown(self):
+        import pyramid.config
+        pyramid.config.global_registries.empty()
+
+    def test_no_header(self):
+        res = self.testapp.get('/', headers={}, status=200)
+        self.assertEqual(res.content_type, 'text/plain')
+
+    def test_header_all(self):
+        res = self.testapp.get('/', headers={'Accept': '*/*'}, status=200)
+        self.assertEqual(res.content_type, 'text/plain')
+
+    def test_header_all_subtypes_of_type(self):
+        res = self.testapp.get('/', headers={'Accept': 'text/*'}, status=200)
+        self.assertEqual(res.content_type, 'text/plain')
+
+    def test_header_specific_media_type(self):
+        res = self.testapp.get(
+            '/', headers={'Accept': 'text/plain'}, status=200,
+        )
+        self.assertEqual(res.content_type, 'text/plain')
+
+    def test_header_ruled_out_by_specific_media_type_q0(self):
+        res = self.testapp.get(
+            '/', headers={'Accept': 'text/plain;q=0, */*'}, status=200,
+        )
+        self.assertEqual(res.content_type, 'text/plain')
 
+    def test_header_ruled_out_by_type_range_q0(self):
+        res = self.testapp.get(
+            '/', headers={'Accept': 'text/*;q=0, text/html'}, status=200,
+        )
+        self.assertEqual(res.content_type, 'text/plain')
+
+    def test_header_ruled_out_by_all_range_q0(self):
+        res = self.testapp.get(
+            '/', headers={'Accept': '*/*;q=0, text/html'}, status=200,
+        )
+        self.assertEqual(res.content_type, 'text/plain')
+
```

Comment:
@mmerickel These tests were written to show the behaviour of Pyramid at the time ([original comment](https://github.com/Pylons/pyramid/pull/3326#issuecomment-414161922)), not what I thought Pyramid should do &mdash; in particular, `test_header_ruled_out_by_specific_media_type_q0`, `test_header_ruled_out_by_type_range_q0`, and `test_header_ruled_out_by_all_range_q0` were intended to show that allowing use of media ranges in the `accept` argument meant that the view would return a content type (`text/plain`) that the client has explicitly said it doesn't accept (`text/plain;q=0, */*`, `text/*;q=0, text/html` and `*/*;q=0, text/html`). You can see this in the same tests for all four test classes &mdash; I was trying to show the inherent  problem in allowing ranges in the `accept` argument.

Reply:
@whiteroses I agree with you that I think those cases should fail and I incorrectly assumed that they were written to show what you thought should be correct, and then we would see which ones passed or failed on various combinations of pyramid/webob. I'd encourage you to submit either fixes to webob or at least updated test cases that demonstrate your intentions.

No comments were added to https://github.com/Pylons/webob/pull/370 so I'm not sure if you're concerned about Pyramid's usage here, or the media range support I added to Webob. I'm assuming that it's Webob itself and if these things are fixed in Webob then the Pyramid tests will start properly failing as you expect.

You should review the state of Webob with respect to https://github.com/Pylons/webob/pull/376.

Reply:
With respect to the q=0 issues, I cannot agree that things are really broken right now. There is no information in the content negotiation that declares that `text/plain` is the content type that's going to be returned. We are not obligated to bend over backward to satisfy the client. In these tests the view is for `*/*` which means it will accept anything - and it does. The only way this behavior could be fixed in Pyramid is by adding some checks in the renderer itself, or in something on egress. As far as ingress is concerned, things seem to work to me and I'm ok with the current assertions in `test_header_ruled_out_by_all_range_q0`, `test_header_ruled_out_by_type_range_q0`, `test_header_ruled_out_by_specific_media_type_q0`.

Reply:
I think my main takeaway from this is that if you want to put any text in the docs that discourages people from using media ranges, I'm fine with that - but I don't think Pyramid is doing the wrong thing here if you do specify media ranges. It's a best-effort match on the range, and if you want to check a specific type then you are responsible for doing that in view code or somewhere else.

Reply:
If you are ok with discouraging people from using media ranges as offers in the docs, then may I ask, *why support media ranges in the first place?* So much of the complexity is just to accommodate media ranges in offers.

I didn't comment on https://github.com/Pylons/webob/pull/370 because I'd written so many times already on the problems with ranges in offers that I thought you knew my view already; I was also hoping that maybe someone else might speak up, or that you might see from the lack of response to [this](https://github.com/Pylons/pyramid/pull/3326#issuecomment-417131147) and [this](https://github.com/Pylons/pyramid/pull/3326#issuecomment-417415052) that nobody seems to care about using media ranges in offers. (We can see, from the PR that originally changed WebOb to accept ranges in offers, that it "Fixes [#155](https://github.com/Pylons/webob/issues/155)". Click through to the WebOb issue, and we can see that @ltvolks himself notes that WebOb not accepting ranges in offers "is not insensible", but conflicts with Pyramid's documentation. Which is what in my view has been happening since: that everyone has been trying to make the behaviour match what was a mistake in the documentation.)

It is important to note, that we've established that media ranges in offers weren't (and aren't) actually working, because of [this](https://github.com/Pylons/pyramid/pull/3326#discussion_r213148720) in Pyramid (as well as the `_check_offer`s in `NilAccept`'s `__contains__` and `best_match` in WebOb).

The problem shown in those three tests above are present not only when the offer is `*/*`, but also when the offer is e.g. `text/*` (as seen in [these three tests](https://github.com/Pylons/pyramid/pull/3326/files#diff-31c901158fb3dca6c7e67ac2216cb057R853)). Yes, "There is no information in the content negotiation that declares that `text/plain` is the content type that's going to be returned" &mdash; but there is only no information when the offer is a range; when the offer is a media type, the offer *is* the content type. Doesn't this seem much more straightforward, and in line with what is described in the RFC? [RFC 7231](https://tools.ietf.org/html/rfc7231#section-5.3.2) is clear on offers being media types: we can see this for example in its discussion of "media type quality factors".

If there are no media ranges in offers, there would be no need for things like `sort_accept_offers`, no need to invent new algorithms around offers, and no need to worry about how these new algorithms interact with what is specified in the RFC, and whether we've thought of everything and accounted for all the cases. It would save you so much work and complexity, from not having to essentially design and maintain an extension to content negotiation as specified by the RFC; and it'd allow us to focus on switching to `acceptable_offers` (which without media ranges in offers should be a lot more straightforward), and dealing with the randomness in accept handling. If you're already at the point where you see the need to discourage people from using media ranges as offers in the docs, could I persuade you to not support media ranges in offers at all?

Reply:
At the very least I cannot talk myself out of supporting ranges in add_route as it is the gateway to more fine-grained content negotiation between views. As far as views go I agree that it’s way simpler to drop support for ranges due to the complexity in `sort_accept_offers` but unfortunately I’m not willing to break compatibility when I see a path forward to maintain it. Which I feel I have accomplished in these PRs. Especially when it’s clear that at least some people were depending on that behavior. I take compatibility seriously. 

I’d break it if I couldn’t support it. But at the end of the day the code works within reason imo. What would you have me do? Remove a bunch of code that I toiled over and have already written to maintain compatibility? I’d still argue to keep the range support in webob for add_route purposes even if we dropped ranges in `sort_accept_offers`.

Finally. `sort_accept_offers` still needs to exist to sort between params and non-params media types in a hierarchy. It just wouldn’t need to deal with the extra level of sorting ranges. It’s not insanely simpler.

One way or another as I said above I think matching against ranges is legitimately useful in add_route.

Reply:
I chatted with @bertjwregeer for a while this morning and we decided that, while things work right now and we could probably ship it, the `q=0` argument with ranges is compelling enough to us that ranges-as-offers probably shouldn't be supported in webob or in add_view. With this being decided, the implementation details of bringing range support to add_route (which I still think is a fine feature) is probably going to be too great and since I also added list-of-offer support to add_route already I'll probably just point people at that instead of supporting ranges there. Finally, I'll probably remove support for ranges in the `order` argument of `sort_accept_offers` (meaning removing ranges from `add_accept_view_order`). This doesn't need to be done, but I think it'll just be annoying to support without being able to rely on some of the parsing code in webob that it currently utilizes.

I'm going to work on making these changes in the next few days.